File: services/backend/src/models/VaultType.ts
import mongoose, { Document, Schema } from "mongoose";

export interface IVaultType extends Document {
    collateralMint: string;
    oraclePriceAccount: string;
    ltvBps: number;
    liqThresholdBps: number;
    liqPenaltyBps: number;
    stabilityFeeBps: number;
    mintFeeBps: number;
    redeemFeeBps: number;
    vaultDebtCeiling: number;
    isActive: boolean;
    totalCollateral: number;
    totalDebt: number;
}

const VaultTypeSchema: Schema = new Schema({
    collateralMint: { type: String, required: true, unique: true },
    oraclePriceAccount: { type: String, required: true },
    ltvBps: { type: Number, required: true },
    liqThresholdBps: { type: Number, required: true },
    liqPenaltyBps: { type: Number, required: true },
    stabilityFeeBps: { type: Number, required: true },
    mintFeeBps: { type: Number, required: true },
    redeemFeeBps: { type: Number, required: true },
    vaultDebtCeiling: { type: Number, required: true },
    isActive: { type: Boolean, required: true, default: true },
    totalCollateral: { type: Number, default: 0 },
    totalDebt: { type: Number, default: 0 },
});

export const VaultType = mongoose.model<IVaultType>("VaultType", VaultTypeSchema);

File: packages/program/programs/aegis_vault/src/state/vault_type.rs
use anchor_lang::prelude::*;

#[account]
pub struct VaultType {
    // Identity
    pub vault_type_id: u32,
    pub collateral_mint: Pubkey,
    pub oracle_price_account: Pubkey,

    // Risk Parameters
    pub ltv_bps: u64,
    pub liq_threshold_bps: u64,
    pub liq_penalty_bps: u64,
    pub stability_fee_bps: u16,
    pub mint_fee_bps: u16,
    pub redeem_fee_bps: u16,

    // Limits
    pub vault_debt_ceiling: u64,
    
    // PDA bumps
    pub vault_authority_bump: u8,
    
    // State
    pub is_active: bool,
    
    // Metadata
    pub created_at: i64,
    pub updated_at: i64,

    // Reserved
    pub reserved: [u8; 96],
}

impl VaultType {
    pub const LEN: usize = 8 + // discriminator
        4 + // vault_type_id
        32 + // collateral_mint
        32 + // oracle_pubkey
        8 + // ltv_bps
        8 + // liq_threshold_bps
        8 + // liq_penalty_bps
        2 + // stability_fee_bps
        2 + // mint_fee_bps
        2 + // redeem_fee_bps
        8 + // vault_debt_ceiling
        1 + // vault_authority_bump
        1 + // is_active
        8 + // created_at
        8 + // updated_at
        96; // reserved
}
